tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,753 @@
+	chkCRC32 := newCRC32()
+
+	// Iterate files in ascending order.
+	seqs := make([]int, 0, len(cdm.mmappedChunkFiles))
+	for seg := range cdm.mmappedChunkFiles {
+		seqs = append(seqs, seg)
+	}
+	sort.Ints(seqs)
+	for _, seq := range seqs {
+		bs := cdm.mmappedChunkFiles[seq]
+		sliceLen := bs.Len()
+		idx := HeadChunkFileHeaderSize
+		for idx < sliceLen {
+			if sliceLen-idx < 30 {
+				return &CorruptionErr{
+					Dir:       cdm.dir.Name(),
+					FileIndex: seq,
+					Err:       errors.Errorf("head chunk file doesn't include enough bytes to read the chunk header - required:%v, available:%v, file:%d", idx+35, sliceLen, seq),
+				}
+			}
+			chkCRC32.Reset()
+			startIdx := idx
+			seriesRef := binary.BigEndian.Uint64(bs.Range(idx, idx+8))
+			idx += 8
+
+			mint := int64(binary.BigEndian.Uint64(bs.Range(idx, idx+8)))
+			idx += 8
+
+			maxt := int64(binary.BigEndian.Uint64(bs.Range(idx, idx+8)))
+			idx += 8
+
+			chunkRef := uint64(seq)<<32 | uint64(idx)
+			if err := f(seriesRef, chunkRef, mint, maxt); err != nil {
+				return &CorruptionErr{
+					Dir:       cdm.dir.Name(),
+					FileIndex: seq,
+					Err:       err,
+				}
+			}
+
+			idx++ // Skip encoding.
+			// Skip the data.
+			dataLen, n := binary.Uvarint(bs.Range(idx, idx+MaxChunkLengthFieldSize))
+			idx += n + int(dataLen)
+
+			// Check CRC.
+			sum := bs.Range(idx, idx+CRCSize)
+			if _, err := chkCRC32.Write(bs.Range(startIdx, idx)); err != nil {
+				return err
+			}
+			if act := chkCRC32.Sum(nil); !bytes.Equal(act, sum) {
+				return errors.Errorf("checksum mismatch expected:%x, actual:%x", sum, act)
+			}
+
+			idx += CRCSize
+		}
+		if idx > sliceLen {
+			// It should be equal to the slice length.
+			return &CorruptionErr{
+				Dir:       cdm.dir.Name(),
+				FileIndex: seq,
+				Err:       errors.Errorf("head chunk file doesn't include enough bytes to read the last chunk data - required:%v, available:%v, file:%d", idx, sliceLen, seq),
+			}
+		}
+	}
+
+	return nil
+}
+
+// Truncate deletes the head chunk files which are strictly below the mint.
+// mint should be in milliseconds.
+func (cdm *ChunkDiskMapper) Truncate(mint int64) error {
+	var removedFiles []int
+
+	cdm.mmappedChunksMtx.RLock()
+	for seq, bs := range cdm.mmappedChunkFiles {
+		if seq == cdm.curFileSequence {
+			continue
+		}
+		b := bs.Range(HeaderMaxtOffset, HeaderMaxtOffset+8)
+		maxt := binary.BigEndian.Uint64(b)
+		if int64(maxt) < mint {
+			removedFiles = append(removedFiles, seq)
+		}
+	}
+	cdm.mmappedChunksMtx.RUnlock()
+
+	return cdm.deleteFiles(removedFiles)
+}
+
+func (cdm *ChunkDiskMapper) deleteFiles(removedFiles []int) error {
+	closers := make([]io.Closer, 0, len(removedFiles))
+	cdm.mmappedChunksMtx.Lock()
+	for _, seq := range removedFiles {
+		cdm.size -= int64(cdm.mmappedChunkFiles[seq].Len())
+		closers = append(closers, cdm.closers[seq])
+		delete(cdm.mmappedChunkFiles, seq)
+		delete(cdm.closers, seq)
+		if err := os.Remove(segmentFile(cdm.dir.Name(), seq)); err != nil {
```

Comment:
I think that's why windows tests fails. You cannot delete before closing.

Reply:
Windows tests failing here are because the machine is not able to handle the mmap size (which hides this one I guess). Good point! I will fix it